File: sdks/ruby/lib/nanoservice/types/node_config.rb
# frozen_string_literal: true

module Nanoservice
  module Types
    # NodeConfig represents node-specific configuration from the runner.
    class NodeConfig
      attr_accessor :name, :path, :node_type, :config

      # @param name [String] Node name (used for registry lookup)
      # @param path [String] Node path in the workflow
      # @param node_type [String] Type of the node
      # @param config [Hash] Key-value configuration for the node
      def initialize(name: "", path: "", node_type: "", config: {})
        @name      = name
        @path      = path
        @node_type = node_type
        @config    = config
      end

      # Build a NodeConfig from a Hash (JSON-parsed).
      # @param hash [Hash] the parsed JSON hash
      # @return [NodeConfig]
      def self.from_hash(hash)
        return new if hash.nil?

        new(
          name:      hash["name"] || "",
          path:      hash["path"] || "",
          node_type: hash["type"] || "",
          config:    hash["config"] || {}
        )
      end

      # Serialize to a Hash suitable for JSON output.
      # @return [Hash]
      def to_hash
        {
          "name"   => @name,
          "path"   => @path,
          "type"   => @node_type,
          "config" => @config
        }
      end

      # Get a string config value with a default.
      # @param key [String] Config key
      # @param default [String] Default value
      # @return [String]
      def config_str(key, default = "")
        val = @config[key]
        val.is_a?(String) ? val : default
      end

      # Get an integer config value with a default.
      # @param key [String] Config key
      # @param default [Integer] Default value
      # @return [Integer]
      def config_int(key, default = 0)
        val = @config[key]
        val.is_a?(Integer) ? val : default
      end

      # Get a boolean config value with a default.
      # @param key [String] Config key
      # @param default [Boolean] Default value
      # @return [Boolean]
      def config_bool(key, default = false)
        val = @config[key]
        val == true || val == false ? val : default
      end
    end
  end
end

File: sdks/php/src/Types/NodeConfig.php
<?php

declare(strict_types=1);

namespace Blok\Blok\Types;

/**
 * NodeConfig represents node-specific configuration from the runner.
 */
final class NodeConfig
{
    public function __construct(
        public string $name = '',
        public string $path = '',
        public string $type = '',
        /** @var array<string, mixed> */
        public array $config = [],
    ) {}

    /**
     * Get a string config value with a default.
     */
    public function configStr(string $key, string $default = ''): string
    {
        $val = $this->config[$key] ?? null;
        return is_string($val) ? $val : $default;
    }

    /**
     * Get an integer config value with a default.
     */
    public function configInt(string $key, int $default = 0): int
    {
        $val = $this->config[$key] ?? null;
        return is_int($val) ? $val : (is_numeric($val) ? (int) $val : $default);
    }

    /**
     * Get a boolean config value with a default.
     */
    public function configBool(string $key, bool $default = false): bool
    {
        $val = $this->config[$key] ?? null;
        return is_bool($val) ? $val : $default;
    }

    /**
     * Create a NodeConfig from an associative array.
     */
    public static function fromArray(array $data): self
    {
        return new self(
            name: $data['name'] ?? '',
            path: $data['path'] ?? '',
            type: $data['type'] ?? '',
            config: $data['config'] ?? [],
        );
    }

    /**
     * Convert to an associative array for JSON serialization.
     */
    public function toArray(): array
    {
        return [
            'name' => $this->name,
            'path' => $this->path,
            'type' => $this->type,
            'config' => $this->config,
        ];
    }
}
